### PR TITLE
Lecture 21: Sequential decision making (part 2): The algorithms

### PR DESCRIPTION
**Pull request by mliu-dark-knight**

### CODE REVIEW

**Comment by Xingyu-Lin** [comment]
rendered version: https://xingyu-lin.github.io/pgm-spring-2019/notes/lecture-21/

File: _posts/2019-04-03-lecture-21.md [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+authors:
+  - name: Xingyu Lin  # author's full name
+    url: "https://xingyu-lin.github.io/"  # optional URL to the author's homepage
+  - name: Mengxiong Liu
+    url: "#"
+  - name: Junyan Jiang
+    url: "#"
+
+editors:
+  - name: Editor 1  # editor's full name
```
Comment (pliang279) [inline _posts/2019-04-03-lecture-21.md]: Paul Liang

File: _posts/2019-04-03-lecture-21.md [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+The approach still involves explicitly optimization of $$\pi_\theta$$. We can rewrite the iteration as:
+$$\begin{array}{l}{\text { 1. set } Q(s, a) \leftarrow r(s, a)+\gamma \mathbb{E}_{s^{\prime} \sim p\left(s^{\prime} | s, a\right)}\left[V\left(s^{\prime}\right)\right]} \\ {\text { 2. } \operatorname{set} V(s) \leftarrow \max _{a} Q(s, a)}\end{array}$$
+
+### Fitted Q-learning:
+
+If the state space is high-dimensional, it is not feasible to represent $$Q$$ as a parametric function. In this case, we adopt fitted Q-iteration:
+
+$$\begin{array}{l}{\text { 1. set } y_{i} \leftarrow r\left(s_{i}, a_{i}\right)+\gamma \mathbb{E}_{s^{\prime} \sim p\left(s^{\prime} | s, a\right)}\left[V_{\phi}\left(s_{i}^{\prime}\right)\right]} \\ {\text { 2. } \operatorname{set} \phi \leftarrow \arg \min _{\phi} \sum_{i}\left\|Q_{\phi}\left(s_{i}, a_{i}\right)-y_{i}\right\|^{2}}\end{array}$$
+
```
Comment (pliang279) [inline _posts/2019-04-03-lecture-21.md:147]: explain more and add links to references

File: _posts/2019-04-03-lecture-21.md [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+<d-math block>
+\begin{aligned}
+J(\theta) &= -D_{KL}(\hat{p}(\tau) || p(\tau)) \\
+&= \sum_{t=1}^T \mathbb{E}_{(s_t, a_t) \sim \hat{p(s_t, a_t)}}[r(s_t, a_t)] + \mathbb{E}_{s_t \sim \hat{p}(s_t)}[\mathcal{H}(\pi(a_t|s_t))]\\
+&= \sum_{t=1}^T \mathbb{E}_{(s_t, a_t) \sim p_\theta(s_t, a_t)}[r(s_t, a_t) - \log \pi(a_t|s_t)]\\
+\end{aligned}​
+</d-math>
+
+Now following the policy gradient method, such as REINFORCE, we just need to add a bonus entropy term to the rewards.
```
Comment (pliang279) [inline _posts/2019-04-03-lecture-21.md:163]: explain more, what is the bonus entropy term?

File: _posts/2019-04-03-lecture-21.md [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+$$
+\theta \gets \theta + \alpha\nabla_\theta Q_\theta(s,a)(r(s,a) + \gamma V(s') - Q_\theta(s,a)),
+$$
+where
+$$
+V(s') = \text{soft}\max_{a'} Q_\theta(s', a') = \log \int \exp (Q_\theta(s', a')) da'.
+$$
+Additionally, we can set the temperature in softmax to control the tradeoff between entropy and rewards.
+
+To summaize, there are a few benefits of soft optimality:
```
Comment (pliang279) [inline _posts/2019-04-03-lecture-21.md:213]: explain in more detail and add references

File: _posts/2019-04-03-lecture-21.md [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+## Types of RL Algorithms
+
+The objective of RL learning is to find the optimal parameter s.t. maximize the expected reward.
+
+$$\theta^*=\arg\max_\theta \mathbb{E}_{\tau\sim p(\tau)}\left[\sum_{t=1}^T r(s_t,a_t)\right]$$
+
+* Policy gradients: directly optimize the above stochastic objective
+* Value-based: estimate V-function or Q-function of the optimal policy(no explicit policy; the policy is derived from the value function)
+* Actor-critic: estimate V-/Q-function under the current policy and use it toimprove the policy (not covered)
```
Comment (pliang279) [inline _posts/2019-04-03-lecture-21.md:104]: toimprove -> to improve

File: _posts/2019-04-03-lecture-21.md [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+* For stochastic dynamics, we obtain it from the ELBO on the evidence.
+
+## Types of RL Algorithms
+
+The objective of RL learning is to find the optimal parameter s.t. maximize the expected reward.
+
+$$\theta^*=\arg\max_\theta \mathbb{E}_{\tau\sim p(\tau)}\left[\sum_{t=1}^T r(s_t,a_t)\right]$$
+
+* Policy gradients: directly optimize the above stochastic objective
+* Value-based: estimate V-function or Q-function of the optimal policy(no explicit policy; the policy is derived from the value function)
```
Comment (pliang279) [inline _posts/2019-04-03-lecture-21.md]: space (no explicit policy

File: _posts/2019-04-03-lecture-21.md [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+
+In policy gradient, we directly optimize the target expected reward w.r.t the policy $\pi_\theta$ itself.
+
+$$\begin{aligned}
+J(\theta)&=\mathbb{E}_{\tau \sim p_{\theta}(\tau)}\left[\sum_{t=1}^{T} r\left(s_{t}, a_{t}\right)\right] \approx \frac{1}{N} \sum_{i=1}^{N} \sum_{t=1}^{T} r\left(s_{i, t}, a_{i, t}\right) \\
+\nabla_{\theta} J(\theta) &=\nabla_{\theta} \mathbb{E}_{\tau \sim p_{\theta}(\tau)}[r(\tau)]=\int r(\tau) \nabla_{\theta} p_{\theta}(\tau) d \tau=\int r(\tau) p_{\theta}(\tau) \nabla_{\theta} \log p_{\theta}(\tau) d \tau \\ &=\mathbb{E}_{\tau \sim p_{\theta}(\tau)}\left[r(\tau) \nabla_{\theta} \log p_{\theta}(\tau)\right] 
+\end{aligned}$$ 
+
+The log-derivative trick is applied to the above equation.
+$$
```
Comment (pliang279) [inline _posts/2019-04-03-lecture-21.md:117]: center these equations

File: _posts/2019-04-03-lecture-21.md [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+\begin{array}{l}{\nabla_{\theta} \log p_{\theta}(\tau)=\nabla_{\theta}\left[\log p\left(s_{1}\right)+\sum_{t=1}^{T} \log p\left(s_{t+1} | s_{t}, a_{t}\right)+\log \pi_{\theta}\left(a_{t} | s_{t}\right)\right]} \\ {\nabla_{\theta} J(\theta)=\mathbb{E}_{\tau \sim p_{\theta}(\tau)}\left[\left(\sum_{t=1}^{T} \nabla_{\theta} \log \pi_{\theta}\left(a_{t} | s_{t}\right)\right)\left(\sum_{t=1}^{T} r\left(s_{t}, a_{t}\right)\right)\right]}\end{array}$$
+
+The reinforce algorithm:
+$$
```
Comment (pliang279) [inline _posts/2019-04-03-lecture-21.md:121]: center these equations

File: _posts/2019-04-03-lecture-21.md [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+### Policy iteration via dynamic programming:
+
+* Policy iteration
+$$\begin{array}{l}{\text { 1. evaluate } Q_{\pi}(s, a)=r(s, a)+\gamma \mathbb{E}_{s^{\prime} \sim p\left(s^{\prime} | s, a\right)}\left[V_{\pi}\left(s^{\prime}\right)\right]} \\ {\text { 2. update } \pi \leftarrow \pi^{\prime}}\end{array}$$
+
+* Policy evaluation
+$$V_{\pi}(s) \leftarrow r(s, \pi(s))+\gamma \mathbb{E}_{s^{\prime} \sim p\left(s^{\prime} | s, \pi(s)\right)}\left[V_{\pi}\left(s^{\prime}\right)\right]$$
+
+The approach still involves explicitly optimization of $$\pi_\theta$$. We can rewrite the iteration as:
```
Comment (pliang279) [inline _posts/2019-04-03-lecture-21.md]: explicitly optimization -> explicit optimization

**Review by pliang279 — COMMENTED** [review]
hi everyone, please address these comments, thanks!

File: _posts/2019-04-03-lecture-21.md [hunk 10]
```diff
@@ -0,0 +1,219 @@
+
```
Comment (alshedivat) [inline _posts/2019-04-03-lecture-21.md]: This empty line mixes up rendering. Please remove.

**Pull request merged**